common/src/chain/config/emission_schedule.rs

```diff
@@ -0,0 +1,611 @@
+
+/// Emission schedule, characterized by function from block height to total supply at that point.
+///
+/// The function has to be a monotonic non-decreasing function.
+#[derive(Clone)]
+pub struct EmissionSchedule(std::sync::Arc<EmissionScheduleFn>);
+
+impl EmissionSchedule {
+    /// Construct an emission schedule from a function.
+    ///
+    /// Be careful to maintain invariants as specified in [EmissionSchedule] docs
+    pub fn from_fn(f: impl 'static + Fn(BlockHeight) -> Mlt + Sync + Send) -> Self {
+        Self(std::sync::Arc::new(f))
+    }
+
+    /// Construct an emission schedule from a function. See also [Self::from_fn]
+    pub fn from_dyn_fn(f: Box<EmissionScheduleFn>) -> Self {
+        Self(f.into())
+    }
+
+    /// Get supply at given block height
+    pub fn supply_at(&self, ht: BlockHeight) -> Mlt {
+        self.0(ht)
+    }
+
+    /// Get initial supply (premine)
+    pub fn initial_supply(&self) -> Mlt {
+        self.supply_at(BlockHeight::zero())
+    }
+
+    /// Get subsidy for block at given height
+    pub fn subsidy(&self, ht: BlockHeight) -> Mlt {
+        let prev_ht = ht.prev_height().expect("Genesis has no subsidy");
+        (self.supply_at(ht) - self.supply_at(prev_ht)).expect("Supply not monotonic")
+    }
+}
+
+impl std::fmt::Debug for EmissionSchedule {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "EmissionSchedule(<function>)")
+    }
+}
+
+/// Emission schedule where supply is a piecewise linear function, represented as a table.
+///
+/// The table has a string representation, described in [Self::from_str].
+#[derive(Eq, PartialEq, Debug, Clone)]
+pub struct EmissionScheduleTabular {
+    /// The initial supply, in MLTs
+    initial_supply: Mlt,
+    /// The initial per-block reward, before the first period kicks in.
+    initial_reward: Mlt,
+    /// Reward periods. Specified by block height and the per-block reward starting at that height.
+    periods: BTreeMap<BlockHeight, Mlt>,
+}
+
+impl EmissionScheduleTabular {
+    /// Create a new piecewise-linear supply schedule
+    ///
+    /// This constructor takes:
+    ///  * The initial supply (premine amount)
+    ///  * The initial per-block reward, before the first explicitly specified period (below).
+    ///  * Reward periods. These are mappings from block heights to per-block rewards from that
+    ///    point onwards, until the next period. The periods are implicitly sorted by BTreeMap.
+    pub fn new(
+        initial_supply: Mlt,
+        initial_reward: Mlt,
+        periods: BTreeMap<BlockHeight, Mlt>,
+    ) -> Self {
+        Self {
+            initial_supply,
+            initial_reward,
+            periods,
+        }
+    }
+
+    /// All reward periods, starting at block height 0
+    pub fn reward_periods(&self) -> impl Iterator<Item = (BlockHeight, Mlt)> + '_ {
+        std::iter::once((BlockHeight::zero(), self.initial_reward))
```

Comment:
Yep... you're manually chaining the initial reward. I don't really see the benefit in complicating the process. Just make element zero the value.

I know you may be doing this to enforce a compile-time check for this, but then there's this trade-off between more code/complexity vs a runtime check. I don't know which is worse.

Reply:
Basically removes a bunch of `expect` and extra invariants placed on the map. The invariant here is not self-contained by this module, so it's basically to ensure that the user of the module cannot violate it by construction.